File: Week 2/Problem Set 2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
void print_grade(string text);

int main(void)
{
    string input = get_string("Text: ");

    print_grade(input);
}

int count_letters(string text)
{
    int countletters = 0;
    int length = strlen(text);

    for (int i = 0; i < length; i++)
    {
        text[i] = tolower(text[i]);

        if (!ispunct(text[i]) && !isdigit(text[i]) && isalpha(text[i]) && (islower(text[i]) || isupper(text[i])))
        {
            countletters ++;
        }
    }
    return countletters;
}

int count_words(string text)
{
    int countwords = 1;
    int length = strlen(text);

    for (int i = 0; i < length; i++)
    {
        if (isspace(text[i]))
        {
            countwords ++;
        }
    }
    return countwords;
}

int count_sentences(string text)
{
    int countsentences = 0;
    int length = strlen(text);

    for (int i = 0; i < length; i++)
    {
        if (text[i] == '!' || text[i] == '.' || text[i] == '?')
        {
            countsentences ++;
        }
    }
    return countsentences;
}

void print_grade(string text)
{
    const int words = 100;

    float average_letters = (((float) count_letters(text) / (float) count_words(text)) * words);
    float average_sentences = (((float) count_sentences(text) / (float) count_words(text)) * words);

    int index = round((0.0588 * average_letters) - (0.296 * average_sentences) - 15.8);

    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

File: Week 6/Problem Set 6/sentimental-readability/readability.py
def count_letters(text):
    countletters = 0

    # Iterate over each character in the text
    for char in text:
        if char.isalpha():
            countletters += 1

    return countletters


def count_words(text):
    # Split the text into individual words using whitespace as the delimiter
    countwords = text.split()

    return len(countwords)


def count_sentences(text):
    # Iterate over each character in the text
    countsentences = 0

    for char in text:
        # Check if the character is a sentence-ending punctuation mark
        if char in ['!', '.', '?']:
            countsentences += 1

    return countsentences


def print_grade(text):
    words = 100

    # Calculate the average number of letters per 100 words
    average_letters = (count_letters(text) / count_words(text)) * words

    # Calculate the average number of sentences per 100 words
    average_sentences = (count_sentences(text) / count_words(text)) * words

    # Calculate the Coleman-Liau index using the average letter and sentence counts
    index = round((0.0588 * average_letters) - (0.296 * average_sentences) - 15.8)

    # Determine the corresponding grade level based on the index
    if index >= 16:
        print("Grade 16+\n")
    elif index < 1:
        print("Before Grade 1\n")
    else:
        print(f"Grade {index}")


input_text = input("Text: ")

print_grade(input_text)